lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -121,6 +123,18 @@ protected Object readArray(Object old, Schema expected, ResolvingDecoder in)
       @SuppressWarnings("unchecked")
       Collection<Object> c = (Collection<Object>) array;
       return readCollection(c, expectedType, l, in);
+    } else if (array instanceof Map 
+               && ReflectData.isMapWithNonStringKeysSchema(expected)) {
```

Comment:
The case is where a subclass of ReflectDatumReader overrides `newArray` and doesn't return an instance of `Map`, but the schema check succeeds. The result is that `array` is treated as a java array, which will likely cause a cast exception or something. Having both checks is what creates this situation. I think the logic should be to check if the schema is a non-string map and then validate that the `array` object can be used as one.

Basically, if the schema indicates it is a map, the right thing to do is validate that the container is a map, rather than moving to the java array case if it isn't.